pkg/cmd/server/origin/master.go

```diff
@@ -131,249 +134,210 @@ import (
+	messages = append(messages, fmt.Sprintf("Started OpenAPI Schema at %%s%s", openAPIServePath))
 
-	handler = open
+	return messages
+}
 
-	// add CORS support
-	if origins := c.ensureCORSAllowedOrigins(); len(origins) != 0 {
-		handler = apiserver.CORS(handler, origins, nil, nil, "true")
+func (c *MasterConfig) buildHandlerChain(assetConfig *AssetConfig) (func(http.Handler, *genericapiserver.Config) (secure, insecure http.Handler), []string, error) {
+	var messages []string
+	if c.Options.OAuthConfig != nil {
+		messages = append(messages, fmt.Sprintf("Started OAuth2 API at %%s%s", OpenShiftOAuthAPIPrefix))
 	}
 
-	if c.WebConsoleEnabled() {
-		handler = assetServerRedirect(handler, c.Options.AssetConfig.PublicURL)
-	}
+	if assetConfig != nil {
+		publicURL, err := url.Parse(assetConfig.Options.PublicURL)
+		if err != nil {
+			return nil, nil, err
+		}
+		messages = append(messages, fmt.Sprintf("Started Web Console %%s%s", publicURL.Path))
+	}
+
+	// TODO(sttts): resync with upstream handler chain and re-use upstream filters as much as possible
+	return func(apiHandler http.Handler, kc *genericapiserver.Config) (secure, insecure http.Handler) {
+		attributeGetter := kapiserverfilters.NewRequestAttributeGetter(c.RequestContextMapper)
+
+		handler := c.versionSkewFilter(apiHandler, c.getRequestContextMapper())
+		handler = c.authorizationFilter(handler)
+		handler = c.impersonationFilter(handler)
+
+		// audit handler must comes before the impersonationFilter to read the original user
+		if c.Options.AuditConfig.Enabled {
+			var writer io.Writer
+			if len(c.Options.AuditConfig.AuditFilePath) > 0 {
+				writer = &lumberjack.Logger{
+					Filename:   c.Options.AuditConfig.AuditFilePath,
+					MaxAge:     c.Options.AuditConfig.MaximumFileRetentionDays,
+					MaxBackups: c.Options.AuditConfig.MaximumRetainedFiles,
+					MaxSize:    c.Options.AuditConfig.MaximumFileSizeMegabytes,
+				}
+			} else {
+				// backwards compatible writer to regular log
+				writer = cmdutil.NewGLogWriterV(0)
+			}
+			handler = kapiserverfilters.WithAudit(handler, attributeGetter, writer)
+		}
+		handler = authenticationHandlerFilter(handler, c.Authenticator, c.getRequestContextMapper())
+		handler = namespacingFilter(handler, c.getRequestContextMapper())
+		handler = cacheControlFilter(handler, "no-store") // protected endpoints should not be cached
 
-	// Make the outermost filter the requestContextMapper to ensure all components share the same context
-	if contextHandler, err := kapi.NewRequestContextFilter(c.getRequestContextMapper(), handler); err != nil {
-		glog.Fatalf("Error setting up request context filter: %v", err)
-	} else {
-		handler = contextHandler
-	}
+		if c.Options.OAuthConfig != nil {
+			authConfig, err := BuildAuthConfig(c)
+			if err != nil {
+				glog.Fatalf("Failed to setup OAuth2: %v", err)
+			}
+			handler, err = authConfig.WithOAuth(handler)
```

Comment:
same here... using a filter to register a handler for a particular set of subpaths seems off to me